internal/blockchain/fabric/fabric.go

```diff
@@ -170,61 +176,113 @@ func (f *Fabric) Init(ctx context.Context, config config.Section, callbacks bloc
+	batchTimeout := uint(f.fabconnectConf.GetDuration(FabconnectConfigBatchTimeout).Milliseconds())
 	if f.initInfo.stream, err = f.streams.ensureEventStream(f.ctx, f.topic, batchSize, batchTimeout); err != nil {
 		return err
 	}
 	log.L(f.ctx).Infof("Event stream: %s", f.initInfo.stream.ID)
-	location := &Location{
-		Channel:   f.defaultChannel,
-		Chaincode: f.chaincode,
-	}
-	if f.initInfo.sub, err = f.streams.ensureSubscription(f.ctx, location, f.initInfo.stream.ID, batchPinEvent); err != nil {
-		return err
-	}
 
 	f.closed = make(chan struct{})
 	go f.eventLoop()
 
 	return nil
 }
 
-func (f *Fabric) Start() error {
+func (f *Fabric) resolveFireFlyContract(ctx context.Context, contractIndex int) (chaincode, fromBlock string, err error) {
+
+	if f.contractConfSize > 0 || contractIndex > 0 {
+		// New config (array of objects under "fireflyContract")
+		if contractIndex >= f.contractConfSize {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgInvalidFireFlyContractIndex, fmt.Sprintf("blockchain.fabric.fireflyContract[%d]", contractIndex))
+		}
+		chaincode = f.contractConf.ArrayEntry(contractIndex).GetString(FireFlyContractChaincode)
+		if chaincode == "" {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "address", "blockchain.fabric.fireflyContract")
+		}
+		fromBlock = f.contractConf.ArrayEntry(contractIndex).GetString(FireFlyContractFromBlock)
+	} else {
+		// Old config (attributes under "ethconnect")
+		chaincode = f.fabconnectConf.GetString(FabconnectConfigChaincodeDeprecated)
+		if chaincode != "" {
+			log.L(ctx).Warnf("The %s.%s config key has been deprecated. Please use %s.%s instead",
+				FabconnectConfigKey, FabconnectConfigChaincodeDeprecated,
+				FireFlyContractConfigKey, FireFlyContractChaincode)
+			fromBlock = string(core.SubOptsFirstEventOldest)
+		} else {
+			return "", "", i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "chaincode", "blockchain.fabric.fabconnect")
+		}
+	}
+
+	return chaincode, fromBlock, nil
+}
+
+func (f *Fabric) ConfigureContract(ctx context.Context, contracts *core.FireFlyContracts) (err error) {
+
+	chaincode, fromBlock, err := f.resolveFireFlyContract(ctx, contracts.Active.Index)
+	if err != nil {
+		return err
+	}
+
+	location := &Location{Channel: f.defaultChannel, Chaincode: chaincode}
+	f.initInfo.sub, err = f.streams.ensureFireFlySubscription(ctx, location, fromBlock, f.initInfo.stream.ID, batchPinEvent)
+	if err == nil {
+		f.fireflyMux.Lock()
+		f.fireflyChaincode = chaincode
+		f.fireflyFromBlock = fromBlock
+		f.fireflyMux.Unlock()
+		contracts.Active.Info = fftypes.JSONObject{
+			"chaincode":    chaincode,
+			"fromBlock":    fromBlock,
+			"subscription": f.initInfo.sub.ID,
+		}
+	}
+	return err
+}
+
+func (f *Fabric) TerminateContract(ctx context.Context, contracts *core.FireFlyContracts, termination *blockchain.Event) (err error) {
+
+	chaincode := termination.Info.GetString("chaincodeId")
+	f.fireflyMux.Lock()
+	if chaincode != f.fireflyChaincode {
```

Comment:
Just an observation about us having added quite a bit of extra code duplication of logic here. At some point it feels like we need a base plugin/framework for the common code between `ethereum` and `fabric`. However, I have no objection to this going in as is - as I'm expecting you were thinking about this duplication as you went through, and how interleaved the logic is.

Reply:
Yea, my gut said it wouldn't improve overall readability to factor out utilities, because so many little details are just slightly different ("address" vs "chaincodeId", for instance). It's probably an exercise worth doing at some point just to see what's possible, but I've punted (again) for now.